src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -1007,9 +1012,54 @@ private PackageLicenseInfo GeneratePackageLicenseInfo(IPackageSearchMetadata met
                 results.Add(result);
             }
 
+            if (newSourceMappings?.Count > 0)
+            {
+                var solutionSourceMappingResult = new PreviewResult(newSourceMappings);
+                results.Add(solutionSourceMappingResult);
+            }
+
             return results;
         }
 
+        private static void GetNewSourceMappingsFromAddedPackages(ref Dictionary<string, SortedSet<string>>? newSourceMappings, UserAction? userAction, List<AccessiblePackageIdentity> added, PackageSourceMapping packageSourceMapping)
+        {
+            string? newMappingSourceName = userAction?.SourceMappingSourceName;
+            if (newMappingSourceName is null || added.Count == 0 || packageSourceMapping is null)
+            {
+                return;
+            }
+
+            List<string> addedPackagesWithNewSourceMappings = added.Select(_ => _.Id)
```

Comment:
The key algorithm of this PR.
If the packages were added by preview restore, but didn't have an existing source mapping, then those mappings are by definition required to be created.